ingestion/src/metadata/ingestion/source/messaging/messaging_service.py

```diff
@@ -0,0 +1,186 @@
+#  Copyright 2021 Collate
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  http://www.apache.org/licenses/LICENSE-2.0
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""
+Base class for ingesting messaging services
+"""
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from typing import Any, Iterable, List, Optional
+
+from confluent_kafka.admin import AdminClient, ConfigResource
+
+from metadata.generated.schema.api.data.createPipeline import CreatePipelineRequest
+from metadata.generated.schema.api.data.createTopic import CreateTopicRequest
+from metadata.generated.schema.entity.data.topic import (
+    SchemaType,
+    Topic,
+    TopicSampleData,
+)
+from metadata.generated.schema.entity.services.connections.metadata.openMetadataConnection import (
+    OpenMetadataConnection,
+)
+from metadata.generated.schema.entity.services.messagingService import (
+    MessagingConnection,
+    MessagingService,
+)
+from metadata.generated.schema.metadataIngestion.messagingServiceMetadataPipeline import (
+    MessagingServiceMetadataPipeline,
+)
+from metadata.generated.schema.metadataIngestion.workflow import (
+    Source as WorkflowSource,
+)
+from metadata.ingestion.api.source import InvalidSourceException, Source, SourceStatus
+from metadata.ingestion.api.topology_runner import TopologyRunnerMixin
+from metadata.ingestion.models.topology import (
+    NodeStage,
+    ServiceTopology,
+    TopologyNode,
+    create_source_context,
+)
+from metadata.ingestion.ometa.ometa_api import OpenMetadata
+from metadata.utils.connections import get_connection, test_connection
+from metadata.utils.filters import filter_by_topic
+
+
+class MessagingServiceTopology(ServiceTopology):
+    """
+    Defines the hierarchy in Messaging Services.
+    service -> messaging -> topics.
+
+    We could have a topology validator. We can only consume
+    data that has been produced by any parent node.
+    """
+
+    root = TopologyNode(
+        producer="get_services",
+        stages=[
+            NodeStage(
+                type_=MessagingService,
+                context="messaging_service",
+                processor="yield_messaging_service",
+            )
+        ],
+        children=["topic"],
+    )
+    topic = TopologyNode(
+        producer="get_topic",
+        stages=[
+            NodeStage(
+                type_=Topic,
+                context="Topic",
+                processor="yield_topic",
+                consumer=["messaging_service"],
+            )
+        ],
+    )
+
+
+@dataclass
```

Comment:
Do you know why we use a dataclass here?

Reply:
`class SourceStatus(Status):` is using dataclass so I am using the same approach.

Reply:
We have a task to overhaul the source status. We'll update to pydantic there https://github.com/open-metadata/OpenMetadata/issues/4694